winpty/src/windows.rs

```diff
@@ -213,26 +244,37 @@ impl Winpty {
     /// before the output data pipe(s) is/are connected, then collected output is
     /// buffered until the pipes are connected, rather than being discarded.
     /// (https://blogs.msdn.microsoft.com/oldnewthing/20110107-00/?p=11803)
-    // TODO: Support getting the process and thread handle of the spawned process (Not the agent)
-    // TODO: Support returning the error from CreateProcess
-    pub fn spawn(&mut self, cfg: &SpawnConfig) -> Result<(), Err> {
+    pub fn spawn(&mut self, cfg: &SpawnConfig) -> Result<ChildHandles, Err> {
+        let mut handles = ChildHandles {
+            process: std::ptr::null_mut(),
+            thread: std::ptr::null_mut()
+        };
+
+        let mut create_process_error: DWORD = 0;
         let mut err = null_mut() as *mut winpty_error_t;
 
         unsafe {
-            let ok = winpty_spawn(
+            winpty_spawn(
                 self.0,
                 cfg.0 as *const winpty_spawn_config_s,
-                null_mut(), // Process handle
-                null_mut(), // Thread handle
-                null_mut(), // Create process error
+                &mut handles.process as *mut _,
+                &mut handles.thread as *mut _,
+                &mut create_process_error as *mut _,
                 &mut err,
             );
-            if ok == 0 {
-                return Ok(());
-            }
         }
 
-        check_err(err).map_or(Ok(()), Result::Err)
+        if let Some(err) = check_err(err) {
+            match err.code {
+                ErrorCode::SpawnCreateProcessFailed => Result::Err(Err {
+                    code: err.code,
+                    message: format!("{} (error code {})", err.message, create_process_error)
+                }),
+                _ => Result::Err(err),
+            }
+        } else {
+            Ok(handles)
+        }
```

Comment:
There are a few ways to rewrite this code to get rid of the nested matching, but I'd like to know for now why we don't always print the full error? Is there any disadvantage to that? I don't see why we would catch just spawnprocess errors with nice messages?

Reply:
If spawning the shell fails, WinPTY sends us back the result of Window's `GetLastError` call. It's called `create_process_error` in this code.

I felt it would be desirable to store this information in the `Err` struct. The simplest way to do that was to add this information to the format string.

In summary, this code is simply forwarding all errors except for `SpawnCreateProcessFailed`, where it's rewriting the `message` contents to add additional information.

Happy to rewrite this if you've got a preferred way the additional information should be stored.

Reply:
My only question is why we wouldn't always add this additional information?

Reply:
> My only question is why we wouldn't always add this additional information?

The `create_process_error` value is only meaningful in the `SpawnCreateProcessFailed` case.

Perhaps a better way is to change `ErrorCode::SpawnCreateProcessFailed` to contain this value?

Reply:
So you're saying that the `check_err` function always adds a message, but that message is never useful unless the error is `SpawnCreateProcessFailed`?

In that case we certainly should add the message only to that specific error, because it would prevent people from using this incorrectly, however that sounds a bit strange to me.

Why would `WINPTY_ERROR_SPAWN_CREATE_PROCESS_FAILED` have an error message, but none of the other errors does?

Reply:
> So you're saying that the check_err function always adds a message, but that message is never useful unless the error is SpawnCreateProcessFailed?

Not quite; All the winpty error codes have a message. (as far as I know)

However, for `SpawnCreateProcessFailed`, winpty also sends the OS error code back from the agent, completely separately to the message. It's this separate error code which is called `create_process_error` above, and only relevant to this error.

I'll modify things to add the code into `SpawnCreateProcessFailed` variant.

Reply:
Ah, I got you now. It's not really possible to add this to `SpawnCreateProcessFailed`, right? Since this is created in the `check_err` method and we wouldn't want to pass an optional field to that method.

Reply:
I probably would have just done the following in that case:

```rust
let result = cherk_err(err);
if let Err(Err { code: ErrorCode::SpawnCreateProcessFailed, message }) = &mut result {
    message = format!("{} (error code {})", message, create_process_error);
}
result
```

Alternatively `map_err` could work out I think, though then we would have a useless `err => err,` in one of the match arms.

I'd also change the return type of `check_err` to `Result<(), Err>`, because every place where we use it we map it through `Result::Err`, which seems mostly pointless.

Reply:
> Ah, I got you now. It's not really possible to add this to SpawnCreateProcessFailed, right? Since this is created in the check_err method and we wouldn't want to pass an optional field to that method.
> 
>Or how would you have achieved this?

I agree that adding an optional field to `check_err` isn't great. I had a couple of ideas brewing:

- Add `check_error_with_os_error`, which takes as an extra argument the `DWORD` value (of `create_process_error`). This would take on the existing `check_err` implementation as well as drop that `DWORD` into the `SpawnCreateProcessFailed` variant. `check_err` would then change to call into that and pass a `DWORD` value of `0`. Still quite messy, but at least all the rest of the functions don't have to write something like `check_err(err, None)`.
- Alternatively, `SpawnCreateProcessFailed` could contain an `Option<DWORD>`. `check_err` would put `None` in there. In the code above we could then put in the value of `create_process_error`.

But I think both of those create a fair amount of mess. So for the moment I'm going to just change the code in the way you suggest.

Reply:
Changing the return type of `check_err` worked nicely. I was able to change the following patterns:

    check_err(err).map_or(Ok(process_list), Result::Err)

into

    check_err(err)?;
    Ok(process_list)

which reads much nicer imo.

Reply:
`
check_err(err)
            .map_or_else(||Ok(handles), |mut e|{
                if e.code == ErrorCode::SpawnCreateProcessFailed {
                    e.message += &format!(" (error code {})", create_process_error);
                }
                Err(e)
            })
`
Too cryptic? I think the add operation on the original message makes the purpose pretty clear, without raising questions about nesting error types...

Reply:
You can also write the following, as a replacement for the `map_or`:
```rust
check_err(err).map(|_| process_list)
```

> Too cryptic? I think the add operation on the original message makes the purpose pretty clear, without raising questions about nesting error types...

I don't think this suggestion has any benefit over the `if let` I've proposed, since you basically just add the `Err(e)` as additional requirement. It's pretty similar to just `map` in this regard.

The `+= &format` isn't really very idiomatic Rust. I've originally had that in my proposal too, but using `format!("{} (error code {})")` is just the far more common strategy in Rust at least as far as I know. If format is used at all, it's usually used to format everything.

There are a few other alternatives, but I think `format!`ing everything is just the best choice at that point.

Reply:
> There are a few other alternatives, but I think format!ing everything is just the best choice at that point.

I agree. I also wondered about `+=`, but I think that'll lead to two allocations: one for `format!` and then another for `+=`. `format!`ing would lead to just one. A micro-optimisation, I know 😄